### PR TITLE
CSR v1 - switch client-go certificate manager utility to v1 by default

### PR DESCRIPTION
**Pull request by liggitt**
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:

* Switches the client-go certificate manager utility (used by kubelet and kubeadm) to use v1 by default, and fall back to v1beta1 if v1 is unavailable

xref https://github.com/kubernetes/enhancements/issues/1513

**Does this PR introduce a user-facing change?**:
```release-note
NONE
```

/milestone v1.19
/priority important-soon
/sig auth
/cc @munnerz @smarterclayton

### CODE REVIEW

File: cmd/kubeadm/app/phases/certs/renewal/apirenewer_test.go [hunk 1]
```diff
@@ -41,18 +43,27 @@ func TestAPIRenewer(t *testing.T) {
...
+	// apirenewer depends on submitting CSRs without signerName, which requires creating via v1beta1.
+	// letting the create use v1beta1 but allowing the list/watch to use v1 requires simulating server-side conversion.
+	// instead, simulate v1 being unavailable so the list/watch falls back to v1beta1.
+	// once the apirenewer create switches to v1, this can be removed.
+	client.PrependReactor("*", "certificatesigningrequests", func(act k8stesting.Action) (bool, runtime.Object, error) {
+		if act.GetResource().Version == "v1" {
+			return true, nil, apierrors.NewNotFound(schema.GroupResource{Group: "certificates.k8s.io", Resource: "certificatesigningrequests"}, "")
+		}
+		return false, nil, nil
+	})
```
Comment (neolit123) [inline cmd/kubeadm/app/phases/certs/renewal/apirenewer_test.go:66]: i think we should just merge https://github.com/kubernetes/kubernetes/pull/90143 before this PR which removes cmd/kubeadm/app/phases/certs/renewal/apirenewer.go

Comment (liggitt) [reply cmd/kubeadm/app/phases/certs/renewal/apirenewer_test.go:66]: sgtm, I'll drop this file if that is in by the time the CSR chain of PRs gets here

File: cmd/kubelet/app/server_bootstrap_test.go [hunk 2]
```diff
@@ -316,7 +316,7 @@ func (s *csrSimulator) ServeHTTP(w http.ResponseWriter, req *http.Request) {
 		}
 		s.csr = csr
 
-	case req.Method == "GET" && req.URL.Path == "/apis/certificates.k8s.io/v1beta1/certificatesigningrequests" && req.URL.RawQuery == "fieldSelector=metadata.name%3Dtest-csr&limit=500&resourceVersion=0":
+	case req.Method == "GET" && req.URL.Path == "/apis/certificates.k8s.io/v1/certificatesigningrequests" && (req.URL.RawQuery == "fieldSelector=metadata.name%3Dtest-csr&limit=500&resourceVersion=0" || req.URL.RawQuery == "fieldSelector=metadata.name%3Dtest-csr"):
```
Comment (deads2k) [inline cmd/kubelet/app/server_bootstrap_test.go:319]: test case change surprises me. I'll keep reading

Comment (liggitt) [reply cmd/kubelet/app/server_bootstrap_test.go:319]: this is a non-informer request trying a v1 list and falling back to a v1beta1 list to see which is available before starting an informer against the right version

File: pkg/kubelet/certificate/bootstrap/bootstrap_test.go [hunk 3]
```diff
@@ -143,59 +177,135 @@ const (
...
+		})
+	default:
+		f.PrependReactor("create", "certificatesigningrequests", func(action clienttesting.Action) (handled bool, ret runtime.Object, err error) {
+			switch action.GetResource().Version {
+			case "v1":
+				if opts.noV1 {
+					return true, nil, apierrors.NewNotFound(certificatesv1.Resource("certificatesigningrequests"), "")
+				}
+				return true, &certificatesv1.CertificateSigningRequest{ObjectMeta: metav1.ObjectMeta{Name: "fake-certificate-signing-request-name", UID: "fake-uid"}}, nil
+			case "v1beta1":
```
Comment (deads2k) [inline pkg/kubelet/certificate/bootstrap/bootstrap_test.go:201]: it's not clear to me why a kubelet would ever create or get a v1beta1 CSR

File: staging/src/k8s.io/client-go/util/certificate/csr/csr.go [hunk 4]
```diff
@@ -27,110 +27,248 @@ import (
...
+	if v1err == nil || !apierrors.IsNotFound(v1err) {
+		return v1req, v1err
+	}
+
+	v1beta1req, v1beta1err := client.CertificatesV1beta1().CertificateSigningRequests().Get(context.TODO(), name, metav1.GetOptions{})
+	if v1beta1err != nil {
+		return nil, v1beta1err
+	}
+
+	v1req = &certificatesv1.CertificateSigningRequest{
```
Comment (deads2k) [inline staging/src/k8s.io/client-go/util/certificate/csr/csr.go:99]: I'm still reading, but a bootleg conversion like this is going to need a test to go with it.

File: staging/src/k8s.io/client-go/util/certificate/csr/csr.go [hunk 5]
```diff
@@ -27,110 +27,248 @@ import (
...
+func create(client clientset.Interface, csr *certificatesv1.CertificateSigningRequest) (reqName string, reqUID types.UID, err error) {
+	// only attempt a create via v1 if we specified signerName and usages and are not using the legacy unknown signerName
+	if len(csr.Spec.Usages) > 0 && len(csr.Spec.SignerName) > 0 && csr.Spec.SignerName != "kubernetes.io/legacy-unknown" {
+		v1req, v1err := client.CertificatesV1().CertificateSigningRequests().Create(context.TODO(), csr, metav1.CreateOptions{})
+		if v1err == nil {
+			return v1req.Name, v1req.UID, v1err
+		}
+	}
+
+	// convert relevant bits to v1beta1
```
Comment (deads2k) [inline staging/src/k8s.io/client-go/util/certificate/csr/csr.go:123]: seems like you should only fall through to this on a `NotFoundError`

File: pkg/kubelet/certificate/bootstrap/bootstrap_test.go (hunk 3 above)
Comment (liggitt) [reply pkg/kubelet/certificate/bootstrap/bootstrap_test.go:201]: yup, dropped the v1beta1 bits from this test

File: staging/src/k8s.io/client-go/util/certificate/csr/csr.go (hunk 4 above)
Comment (liggitt) [reply staging/src/k8s.io/client-go/util/certificate/csr/csr.go:99]: if we hit this path, we're talking to a pre-1.19 server, which wouldn't have the new fields :-)

Comment (deads2k) [reply staging/src/k8s.io/client-go/util/certificate/csr/csr.go:99]: > if we hit this path, we're talking to a pre-1.19 server, which wouldn't have the new fields :-)

Sneaky.  I think technically I could turn off v1 csr intentionally, but I'll agree that's unlikely.

File: staging/src/k8s.io/client-go/util/certificate/csr/csr.go (hunk 5 above)
Comment (liggitt) [reply staging/src/k8s.io/client-go/util/certificate/csr/csr.go:123]: done

File: staging/src/k8s.io/client-go/util/certificate/csr/csr.go (hunk 4 above)
Comment (deads2k) [reply staging/src/k8s.io/client-go/util/certificate/csr/csr.go:99]: > if we hit this path, we're talking to a pre-1.19 server, which wouldn't have the new fields :-)

That assumes defaults won't have value in this code.

Comment (liggitt) [reply staging/src/k8s.io/client-go/util/certificate/csr/csr.go:99]: The same would apply to future fields added and defaulted in v1... if we start using new fields and depending on server-side defaults, we'll need a fallback when the field is absent from a v1 object read from a 1.19 server, not just in the conversion case

**Comment by liggitt** [comment]
unrelated services e2e failures
/retest

**Comment by liggitt** [comment]
verify timeout
/retest

**Comment by liggitt** [comment]
rebased test/e2e/framework/.import-restrictions

**Pull request merged**